### PR TITLE
perf(fibre): flat-file shard store for fibre

### PR DESCRIPTION
**Pull request by walldiss**
## Problem

Pebble's commit pipeline serializes large-value writes through a single goroutine, so 28 MiB shard `store.Put` calls at high concurrency become the upload-path bottleneck. Measured on a 9v × 10e × c=2 talis cluster (c6id.8xlarge):

| Disk | `store_put` avg | aggregate upload |
|---|---|---|
| EBS gp3 (root) | 12.26 s | 213 MB/s |
| Local NVMe | 0.32 s | 1.10 GB/s |

Even on fast disk, pebble is the limiter. At c=4 per-Put time *grows* to 1.78 s on the same NVMe — pebble's batch-commit serialization gets *worse* as concurrency rises.

## Solution

Move the 28 MiB shard payload off pebble. It now lives as a flat file at `<store>/shards/<commitment>-<promise-hash>` written via a custom binary codec (no `proto.Marshal` of the bulk data). Pebble retains only the small metadata:

- promise (`/pp/<hash>`)
- shard marker (`/shard/<commitment>/<hash>` with empty value, used to iterate by commitment)
- prune index (`/prune/<ts>/<commitment>/<hash>`)

Writes follow stage → commit → publish:

1. The shard is staged to a uniquely-named file under `<store>/staging/`.
2. Pebble metadata is committed in a single batch.
3. The tmp file is renamed to its canonical path, atomically publishing the shard.

The store is filesystem-agnostic via `pebble/v2/vfs.FS`; production uses `vfs.Default`, and `NewMemoryStore` uses `vfs.NewMem` so both pebble metadata and shard files live entirely in memory (no on-disk tempdir, nothing to clean up on `Close`).

## Wire format (on-disk)

Length-prefixed, all big-endian:

```
uint32  version
uint32  root_len + root
uint32  coeffs_len + coefficients
uint32  num_rows
  for each row:
    uint32  index
    uint32  data_len + data
    uint32  num_proof_segments
    for each segment: uint32 seg_len + seg
```

Capped at 1 GiB per byte-length, 65536 rows, 64 proof segments per row, so corrupt files are rejected before any large allocation.

## Crash recovery

- **Process killed before rename**: stale tmp under `staging/` → dropped on next open by `Store.reconcile` (one `RemoveAll`).
- **Rename succeeded, pebble lost the WAL on power loss** (`pebble.NoSync`): orphan file with no marker. Bounded leak (~28 MiB per power-loss event); operator can sweep manually if it ever matters.
- **Rename failed after pebble commit**: phantom marker with no file. `Store.Get` deletes the marker inline on first read; `PruneBefore` cleans the rest at `pruneAt`.

## Measured impact

c6in.8xlarge + provisioned gp3 (1 GB/s, 16K IOPS), c=4 single encoder:

| Stage | proto + pebble | flat-file | Δ |
|---|---|---|---|
| `server_store_put` avg | 1.78 s | **0.033 s** | **54× faster** |
| `server_handler` avg | 1.84 s | 0.09 s | 20× faster |
| aggregate upload data | 1.20 GB/s | **2.13 GB/s** | +78% |

## Compatibility

**Breaking on-disk format.** Existing pebble-stored shards are not migrated — operators wipe the entire fibre store directory before running this version. Acceptable for fibre testnets and fresh deployments. Upgrades on production validators with live shard data would need a migration tool (out of scope here).

`NewPebbleStore` is renamed to `NewStore` since the store is no longer pebble-only (pebble metadata + flat files); `ServerConfig.StoreFn` defaults updated.

Closes: https://linear.app/celestia/issue/PROTOCO-1613/fibre-flat-file-shard-store-bypass-pebble-for-28-mib-shard-payloads

## Test plan

- [x] Full `fibre/...` suite passes, including 13 store tests (Put/Get round-trip, deterministic ordering, prune-by-time, RLC coefficient preservation, concurrent same-key puts, multi-promise edge cases, lazy orphan-marker cleanup, identical-pruneAt sweep) and 7 codec tests (round-trip, parser-bomb rejection, mid-row truncation).
- [x] `go test -race` passes on the store package.
- [x] `make lint` — golangci-lint passes.
- [x] Manual run on talis: 9v × 1e × c=20 sustained 1 GB/s data with `store_put` averaging 18 ms.

---

<a href="https://app.devin.ai/review/celestiaorg/celestia-app/pull/7190" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: fibre/server.go [hunk 1]
```diff
@@ -100,7 +100,11 @@ func (s *Server) Start(ctx context.Context) (err error) {
 	}
 	s.log.Info("signer ready")
 
-	s.store, err = s.Config.StoreFn(s.Config.StoreConfig)
+	storeCfg := s.Config.StoreConfig
+	if storeCfg.Log == nil {
+		storeCfg.Log = s.log
+	}
```
Comment (Wondertan) [inline fibre/server.go]: nit: set the default logger in the default config function instead.

File: fibre/store.go [hunk 2]
```diff
@@ -41,116 +63,175 @@ func (cfg StoreConfig) Validate() error {
...
+	tmp, err := s.writeTmpShard(shard)
+	if err != nil {
+		return fmt.Errorf("writing shard tmp: %w", err)
+	}
+	finalPath := s.shardFilePath(promise.Commitment, promiseHash)
 
-	// write payment promise
 	promiseProto, err := promise.ToProto()
 	if err != nil {
+		_ = s.fs.Remove(tmp)
```
Comment (Wondertan) [inline fibre/store.go]: nit: extract everything after the shard write into a method, so you don't have to do this removal after every erroneous call, but one.

File: fibre/store.go [hunk 3]
```diff
@@ -271,13 +362,48 @@ func (s *Store) PruneBefore(_ context.Context, before time.Time) (int, error) {
 	return pruned, nil
 }
 
-// Close closes the underlying Pebble database.
+// reconcile drops everything under <store>/staging/. Anything there at open
+// time is a leftover from a Put that crashed before the rename. Orphan
+// markers and orphan files in shards/ are intentionally not cleaned here:
+// markers self-heal in [Store.Get] and at pruneAt via [Store.PruneBefore];
+// rare orphan files (pebble.NoSync power loss after rename) are accepted.
+func (s *Store) reconcile() error {
```
Comment (Wondertan) [inline fibre/store.go:378]: You can also write to the temp folder every OS has (/tmp in linux). Then the OS will cleanup things for you without needing a custom reconcile

File: fibre/store.go [hunk 4]
```diff
@@ -41,116 +63,175 @@ func (cfg StoreConfig) Validate() error {
...
+// Get returns the first [types.BlobShard] found for the given [Commitment].
+// When multiple promises exist for the same commitment, returning only the
+// first prevents unbounded message sizes; pebble's deterministic key order
+// makes the choice consistent across validators.
 //
-// If unmarshaling fails for some entries, it continues trying others.
-// Returns an error only if all entries fail to unmarshal or if no shards are found.
+// Get may write to pebble: if a /shard/ marker is found but the backing file
+// is missing (crash leftover or pebble.NoSync power loss), the marker is
+// deleted inline so future Gets stop paying the missed lookup.
```
Comment (Wondertan) [inline fibre/store.go:242]: How is this possible if we write the file, and only after updating the index in pebble? AFAIU, if there is value, there has to be a file, even in crash or power-loss cases.

File: fibre/store.go (hunk 3 above)
Comment (Wondertan) [reply fibre/store.go:378]: os.TempDir would do the thing

File: fibre/store_codec.go [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+//	  for each row:
+//	    uint32 index
+//	    uint32 data_len; []byte data
+//	    uint32 num_proof_segments
+//	      for each: uint32 segment_len; []byte segment
+//
+// Replaces gogoproto.Marshal of BlobShard to avoid one 28 MiB allocation +
+// memcpy per Put — the proto path dominated the encoder CPU profile
+// (memmove + memclr ≈ 39%) at high concurrency.
+const shardCodecVersion uint32 = 1
```
Comment (Wondertan) [inline fibre/store_codec.go:32]: I am really surprised there is no proto.Write(w, msg) or smlr that would write any proto message without having to allocate a new buffer. 

Then I recalled libp2p has https://github.com/libp2p/go-msgio/blob/master/protoio/uvarint_reader.go, which now explains why they added it.

Tbf, I would prefer importing this(or even copying) instead of writing an entire binary format.

P.S I think we should drop the proto eventually tho as per #6805, but thats together with GRPC

File: fibre/store.go (hunk 4 above)
Comment (walldiss) [reply fibre/store.go:242]: there are 3 steps:
1. write to tmp file
2. write to pebble
3. rename file

If crash happens between 2 and 3 there might be value in pebble, but no file in `/shards` dir.

**Review by evan-forbes — DISMISSED** [review]
the code looks straightforward enough, however that kinda worries me that it could be too specific and potentially miss weird edge cases

can be a flup, but its always nice to have more round trip fuzzing even if we're the ones writing and reading here.

It would be weird if someone could actually cause a panic here since we already have verification etc, but its cheap to add and adds some level of comfort for me at least

File: fibre/server.go (hunk 1 above)
Comment (walldiss) [reply fibre/server.go]: Done — moved into `ServerConfig.Validate` right after the `cfg.Log` default, so `Start` is back to a single line. 3f7bf89

File: fibre/store.go (hunk 2 above)
Comment (walldiss) [reply fibre/store.go]: Done — pulled the pebble batch + rename into `commitAndPublish`; `Put` has one `_ = s.fs.Remove(tmp)` site now. 3f7bf89

File: fibre/store.go (hunk 3 above)
Comment (walldiss) [reply fibre/store.go:378]: Going to stick with `<store>/staging/`. `os.Rename` errors with `EXDEV` across filesystems (`/tmp` is often tmpfs or a separate volume) and would force a copy-then-unlink fallback — exactly the extra memcpy this PR exists to avoid. It also breaks the `vfs.FS` abstraction `NewMemoryStore` rides on. `reconcile` is ~10 lines, seems worth keeping for the symmetry.

File: fibre/store_codec.go (hunk 5 above)
Comment (walldiss) [reply fibre/store_codec.go:32]: `protoio` / `proto.Write` still marshals each message body into a buffer before writing the length prefix — that 28 MiB allocation per Put is the bottleneck the codec exists to skip. Per-row marshaling is a theoretical middle ground but keeps gogoproto on the hot path. Once #6805 lands the codec is already shaped right. Added a comment in the file calling this out, and added round-trip + read-no-panic fuzzers per @evan-forbes. 3f7bf89

Comment (Wondertan) [reply fibre/store_codec.go:32]: That buffer is reusable, so it's a single allocation only for the entirety of the process.

Comment (Wondertan) [reply fibre/store_codec.go:32]: Although it's not that easy to reuse, given that we don't have fixed write workers.

Comment (Wondertan) [reply fibre/store_codec.go:32]: Ok, let's keep it then. I am not reviewing the codec itself yet, given its current use only, and the actual networking codec will likely need a fresh look when it comes to it.

File: fibre/store.go [hunk 6]
```diff
@@ -41,75 +63,98 @@ func (cfg StoreConfig) Validate() error {
...
 	}
 
-	return &Store{
-		cfg: cfg,
-		db:  db,
-	}, nil
+	log := cfg.Log
+	if log == nil {
+		log = slog.Default()
+	}
```
Comment (Wondertan) [inline fibre/store.go]: nit: put in Validate?

File: fibre/store.go [hunk 7]
```diff
@@ -118,39 +163,78 @@ func (s *Store) Put(_ context.Context, promise *PaymentPromise, shard *types.Blo
...
 	}
-	if err := batch.Set(shardKey(promise.Commitment, promiseHash), shardData, pebbledb.NoSync); err != nil {
-		return fmt.Errorf("putting shard: %w", err)
+	return nil
+}
+
+// writeTmpShard stages a shard under <store>/staging/ at a randomly named
+// file. Random (not canonical staging/<commit>-<hash>) because vfs.FS.Create
+// truncates on collision rather than failing — no O_EXCL — so two concurrent
+// same-key writers would clobber each other's tmp. Random per-writer names
```
Comment (Wondertan) [inline fibre/store.go:197]: when can we have two concurrent same key writers?

File: fibre/store_test.go [hunk 8]
```diff
@@ -294,11 +349,145 @@ func testStoreGetDeterministicOrdering(t *testing.T, store *fibre.Store) {
 	}
 }
 
+// Reconcile drops staging/ leftovers on open, leaves real shards alone, and
+// logs the cleanup count.
+func TestStoreReconcileStaging(t *testing.T) {
```
Comment (Wondertan) [inline fibre/store_test.go:357]: nit: all the existing tests are grouped together and this and  the ones below could be too for consistency

**Review by Wondertan — DISMISSED** [review]
mainly a last question

File: fibre/store_codec.go (hunk 5 above)
Comment (walldiss) [reply fibre/store_codec.go:32]: Thanks, agreed. Happy to revisit the codec when the networking path lands.

File: fibre/store.go (hunk 6 above)
Comment (walldiss) [reply fibre/store.go]: Done. Flipped `StoreConfig.Validate` to a pointer receiver, moved the `slog.Default()` fallback there, and `openStore` now calls `cfg.Validate()` up front. Matches the `ServerConfig.Validate` pattern. 218cf5a4d

File: fibre/store.go (hunk 7 above)
Comment (walldiss) [reply fibre/store.go:197]: Practical trigger is an upload retry: the uploader (or tx-injector) resends the same payment promise while the first `UploadShard` is still in flight, so two goroutines can land in `Put` for the same `(commitment, promiseHash)`. We don't dedupe at the RPC layer, and a client can also do it deliberately. `testStorePutConcurrentSameKey` pins the property at N=50.

Even though `os.Rename` is atomic-replace and the verified shard bytes are deterministic (so last-writer-wins is a no-op for correctness), I'd rather keep the store contract "concurrent same-key `Put` is safe" instead of relying on callers serializing. Random tmp names give that for free without per-key locking: each writer owns its own tmp and `Rename` is the one publish step. Canonical staging names would race the writes themselves (`vfs.Default.Create` would unlink-and-recreate, `MemFS.Create` would stomp the dirent), so making that path safe would need a per-key mutex, which is strictly more code than the random suffix.

If we also want to kill the wasted `verifyShard` CPU on duplicate retries, that's a fast-path dedupe at `UploadShard` (`store.GetPaymentPromise(hash)` check before `verifyShard`), separate from the store change. Happy to open a follow-up issue for it.

File: fibre/store_test.go (hunk 8 above)
Comment (walldiss) [reply fibre/store_test.go:357]: Done. Folded the three orphan-marker tests (`Get_CleansOrphanMarker`, `Get_SkipsOrphanToSibling`, `Get_AllOrphans`) into the `TestStore` table; harness now passes `(store, path)` so they can reach the shard files. Kept `TestStoreReconcileStaging` standalone because it specifically exercises open-time behavior (close + reopen with a custom log buffer to assert the cleanup count log line), which doesn't fit a shared-store harness. 218cf5a4d

File: fibre/store.go (hunk 7 above)
Comment (Wondertan) [reply fibre/store.go:197]: Def worth parking into an issue

**Review by Wondertan — APPROVED** [review]
